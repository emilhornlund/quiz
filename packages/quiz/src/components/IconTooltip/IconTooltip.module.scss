@import '../../styles/colors.scss';
@import '../../styles/helpers.scss';

.iconTooltipButton {
  -moz-appearance: none;
  -o-appearance: none;
  -webkit-appearance: none;
  appearance: none;
  background-color: transparent;
  border: none;
  font-family: inherit;
  outline: none;

  @include mobile {
    @include fontSize($mobile-control-normal-font-size);
  }

  @include tablet {
    @include fontSize($tablet-control-normal-font-size);
  }

  @include desktop {
    @include fontSize($desktop-control-normal-font-size);
  }
}

.iconTooltip {
  display: flex;
  flex-direction: column;
  color: $gray-4;
  z-index: $tooltip-z-index;
  background-color: $turquoise-1;
  max-width: $mobile-width;

  @include mobile {
    @include fontSize($mobile-control-normal-font-size);

    row-gap: $mobile-narrow-layout-gutter * 0.5;
    padding: $mobile-narrow-layout-gutter * 0.5 $mobile-narrow-layout-gutter;
    border-radius: $mobile-control-small-border-radius;
  }

  @include tablet {
    @include fontSize($tablet-control-normal-font-size);

    row-gap: $tablet-narrow-layout-gutter * 0.5;
    padding: $tablet-narrow-layout-gutter * 0.5 $tablet-narrow-layout-gutter;
    border-radius: $tablet-control-small-border-radius;
  }

  @include desktop {
    @include fontSize($desktop-control-normal-font-size);

    row-gap: $desktop-narrow-layout-gutter * 0.5;
    padding: $desktop-narrow-layout-gutter * 0.5 $desktop-narrow-layout-gutter;
    border-radius: $desktop-control-small-border-radius;
  }

  ul {
    margin: 0;
  }
}
