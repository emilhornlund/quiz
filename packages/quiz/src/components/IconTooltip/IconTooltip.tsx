import {
  autoUpdate,
  flip,
  offset,
  shift,
  useDismiss,
  useFloating,
  useFocus,
  useHover,
  useInteractions,
  useRole,
} from '@floating-ui/react'
import { IconDefinition } from '@fortawesome/fontawesome-common-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import React, { FC, ReactNode, useState } from 'react'

import styles from './IconTooltip.module.scss'

export interface IconTooltipProps {
  icon: IconDefinition
  children: ReactNode | ReactNode[] | string
}

const IconTooltip: FC<IconTooltipProps> = ({ icon, children }) => {
  const [isOpen, setIsOpen] = useState(false)

  const { refs, floatingStyles, context } = useFloating({
    open: isOpen,
    onOpenChange: setIsOpen,
    middleware: [offset(10), flip(), shift()],
    whileElementsMounted: autoUpdate,
  })

  const hover = useHover(context, { move: false })
  const focus = useFocus(context)
  const dismiss = useDismiss(context)
  const role = useRole(context, {
    role: 'label',
  })

  const { getReferenceProps, getFloatingProps } = useInteractions([
    hover,
    focus,
    dismiss,
    role,
  ])

  return (
    <>
      <button
        ref={refs.setReference}
        className={styles.iconTooltipButton}
        {...getReferenceProps()}>
        <FontAwesomeIcon icon={icon} />
      </button>
      {isOpen && (
        <div
          ref={refs.setFloating}
          style={floatingStyles}
          className={styles.iconTooltip}
          {...getFloatingProps()}>
          {children}
        </div>
      )}
    </>
  )
}

export default IconTooltip
