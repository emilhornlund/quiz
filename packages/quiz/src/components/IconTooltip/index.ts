export type { IconTooltipProps } from './IconTooltip'
export { default } from './IconTooltip'
