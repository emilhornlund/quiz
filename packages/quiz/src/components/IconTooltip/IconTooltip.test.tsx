import { faCircleInfo } from '@fortawesome/free-solid-svg-icons'
import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import IconTooltip from './IconTooltip'

describe('IconTooltip', () => {
  it('should render IconTooltip with default props', () => {
    const { container } = render(
      <IconTooltip icon={faCircleInfo}>Tooltip text</IconTooltip>,
    )

    expect(container).toMatchSnapshot()
  })
})
