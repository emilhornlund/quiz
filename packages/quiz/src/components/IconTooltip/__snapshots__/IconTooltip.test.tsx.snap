// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IconTooltip > should render IconTooltip with default props 1`] = `
<div>
  <button
    class="iconTooltipButton"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-info "
      data-icon="circle-info"
      data-prefix="fas"
      role="img"
      viewBox="0 0 512 512"
    >
      <path
        d="M256 512a256 256 0 1 0 0-512 256 256 0 1 0 0 512zM224 160a32 32 0 1 1 64 0 32 32 0 1 1 -64 0zm-8 64l48 0c13.3 0 24 10.7 24 24l0 88 8 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-80 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l24 0 0-64-24 0c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
